restic/commands.go

```diff
@@ -0,0 +1,575 @@
+		return parseStream(file, commandName)
+	} else {
+		return nil, err
+	}
+}
+
+func tryParseVersion(version string) *semver.Version {
+	if v, err := semver.NewVersion(version); err != nil {
+		clog.Tracef("failed parsing restic version %q: %s", version, err.Error())
+		return nil
+	} else {
+		return v
+	}
+}
+
+// ParseCommandsFromManPages parses commands from manual pages located in the specified directory
+// and adds them to the known commands. version indicates the restic version that generated the man pages and
+// baseVersion indicates that the version is the min required base version and will not be used for filtering
+func ParseCommandsFromManPages(manualDir fs.FS, version string, baseVersion bool) error {
+	if commands == nil {
+		commands = map[string]*command{}
+	}
+	return parseCommandsFromManPagesInto(manualDir, version, baseVersion, commands)
+}
+
+func parseCommandsFromManPagesInto(manualDir fs.FS, version string, baseVersion bool, commands map[string]*command) error {
+	filePattern := regexp.MustCompile("^restic(|-.+)\\.1$")
+	newCommands := map[string]*command{}
+
+	if files, err := fs.ReadDir(manualDir, "."); err == nil {
+		for _, file := range files {
+			if m := filePattern.FindStringSubmatch(file.Name()); m != nil && !file.IsDir() {
+				var cmd *command
+				if cmd, err = parseFile(manualDir, file.Name(), m[1]); err != nil {
+					return err
+				}
+
+				if cmd != nil {
+					cmd.FromVersion = version
+					for i := range cmd.Options {
+						cmd.Options[i].FromVersion = version
+					}
+
+					newCommands[cmd.Name] = cmd
+				}
+			}
+		}
+	} else {
+		return err
+	}
+
+	base := ""
+	if baseVersion {
+		base = version
+	}
+	mergeCommandsInto(base, newCommands, commands)
+
+	return nil
+}
+
+func mergeCommandsInto(baseVersion string, source, target map[string]*command) {
+	sourceVersion := ""
+
+	for _, sc := range source {
+		sourceVersion = sc.FromVersion
+
+		if tc := target[sc.Name]; tc == nil || len(tc.Options) == 0 {
+			// Add new command
+			sc.sortOptions()
+			target[sc.Name] = sc
+
+		} else {
+			if len(tc.Description) < len(sc.Description) {
+				tc.Description = sc.Description
+			}
+
+			// Merge options
+			optionsSet := func(options []Option) (opts map[string]int) {
+				opts = make(map[string]int)
+				for i := range options {
+					opts[options[i].Name] = i
+				}
+				return
+			}
+
+			// add new options
+			targetOpts := optionsSet(tc.Options)
+			for _, sopt := range sc.Options {
+				if i, exists := targetOpts[sopt.Name]; exists {
+					tc.Options[i].Alias = sopt.Alias
+					tc.Options[i].Default = sopt.Default
+					tc.Options[i].Description = sopt.Description
+					tc.Options[i].Once = sopt.Once
+				} else {
+					tc.Options = append(tc.Options, sopt)
+				}
+			}
+
+			// record removed options
+			sourceOpts := optionsSet(sc.Options)
+			for i := range tc.Options {
+				if _, exists := sourceOpts[tc.Options[i].Name]; !exists {
+					tc.Options[i].RemovedInVersion = sourceVersion
```

Comment:
Should we only set `RemovedInVersion` when it's empty? Otherwise it will look like options are always removed at the latest know version 😜

Reply:
Very good catch. This fixes version information for 2 flags. Will commit the change when I have it unit-tested.